tripal/tripal
# Support drupal fields by adding a DrupalSQLStorage implementation

Issue #1611 
Previous PR #1612 

# Tripal Version: 4

# Description

This PR adds an implementation of TripalStorage called DrupalSqlStorage which is focused on providing support for Tripal fields using the Drupal storage backend. Tbh DrupalSqlStorage doesn't actually do anything but rather stands in as a placeholder in case we need more specialized functionality at a later date. It's presence does allow us to create Tripal fields and specify the storage as drupal_sql_storage and then TripalStorage will be happy :-) and the referenced bug is solved!

I also added in a lot of automated testing for Tripal Storage as it seemed as good a time as any and wasn't too much work :-)

### Small Bugs found during testing and fixed

- PropertyStorageBase did not check that required values were actually provided.
- TripalStorageBase::addFieldDefinition() needlessly assigned an array the first time a field was added and then promptly overrode it with the definition
- TripalStorageUpdateException was never used so I removed it.
- VarCharStoragePropertyType did not ensure the max size was an int. This allowed people to accidentally pass in the storage settings instead 🙈 
- add/get/remove Types functionality should have been in TripalStorageBase.
- An error should be thrown if a field type is added without storage indicated.

## Testing?

1. Create a fresh docker image + container based on this branch ([see these instructions if needed](https://tripaldoc.readthedocs.io/en/latest/contributing/docker-for-testing.html#testing-an-unmerged-pull-request)).
2. Copy the following field class files into the tripal/src/Plugin/Field/FieldType, FieldWidget and FieldFormatter folders and replace `.txt` with `.php`: 
[TripalExampleSimpleStringTypeItem.txt](https://github.com/tripal/tripal/files/12528965/TripalExampleSimpleStringTypeItem.txt), [TripalExampleSimpleStringWidget.txt](https://github.com/tripal/tripal/files/12388874/TripalExampleSimpleStringWidget.txt),  [TripalExampleSimpleStringFormatter.txt](https://github.com/tripal/tripal/files/12388875/TripalExampleSimpleStringFormatter.txt)
3. Rebuild cache `drush cr`
4. Go to Tripal > Page Structure > Organism > Manage Fields and click on "+ Create a new field"
5. Select "Tripal Example Simple String Field Type" from the "Add a new field" drop-down. Enter a label for the field and click "Save and Continue".
6. Use the default field settings and click Save. Set a Term and click Save.
7. Now create an organism by going to Admin > Tripal > Content and clicking "Add Tripal Content". This branch should show you a form element for the new field you added with no errors. Fill out this field and all required fields and click save.
8. The view page should show the value of the field just as you entered it.
9. Edit the page and change the field value then click save. It should be updated as expected with no errors.
